ecl/hthor/hthor.cpp

```diff
@@ -10363,6 +10364,681 @@ void CHThorExternalActivity::stop()
+}
+
+bool CHThorNewDiskReadBaseActivity::openFirstPart()
+{
+    partNum = 0;
+    if (dfsParts)       // more should really be fileDesc or something
+    {
+        if (dfsParts->first())
+        {
+            openFilePart(ldFile, &dfsParts->query(), 0);
+            return true;
+        }
+    }
+    else if (ldFile)
+    {
+        if (ldFile->numParts() != 0)
+        {
+            openFilePart(ldFile, nullptr, 0);
+            return true;
+        }
+    }
+    else if (!tempFileName.isEmpty())
+    {
+        openFilePart(tempFileName);
+        return true;
+    }
+    setEmptyStream();
+    return false;
+}
+
+bool CHThorNewDiskReadBaseActivity::openNextPart()
+{
+    if (finishedParts)
+        return false;
+
+    offset_t sizeFilePart = 0;
+    if (dfsParts)
+        sizeFilePart = dfsParts->query().getFileSize(true, false);
+    else if (ldFile)
+        sizeFilePart = ldFile->getPartFileSize(partNum);
+
+    offsetOfPart += sizeFilePart;
+    closepart();
+
+    partNum++;
+    if (dfsParts)
+    {
+        if (dfsParts->next())
+        {
+            openFilePart(ldFile, &dfsParts->query(), partNum);
+            return true;
+        }
+    }
+    else if (ldFile)
+    {
+        if (partNum < ldFile->numParts())
+        {
+            openFilePart(ldFile, nullptr, partNum);
+            return true;
+        }
+    }
+    setEmptyStream();
+    return false;
+}
+
+bool CHThorNewDiskReadBaseActivity::initStream(IDiskRowReader * reader, const char * filename)
+{
+    activeReader = reader;
+    if (useRawStream)
+        rawRowStream = reader->queryRawRowStream();
+    else
+        roxieRowStream = reader->queryAllocatedRowStream(rowAllocator);
+
+    StringBuffer report("Reading file ");
+    report.append(filename);
+    agent.reportProgress(report.str());
+
+    return true;
+}
+
+void CHThorNewDiskReadBaseActivity::setEmptyStream()
+{
+    if (useRawStream)
+        rawRowStream = queryNullRawRowStream();
+    else
+        roxieRowStream = queryNullAllocatedRowStream();
+    finishedParts = true;
+}
+
+IDiskRowReader * CHThorNewDiskReadBaseActivity::ensureRowReader(const char * format, bool streamRemote, unsigned expectedCrc, IOutputMetaData & expected, unsigned projectedCrc, IOutputMetaData & projected, unsigned actualCrc, IOutputMetaData & actual, IPropertyTree * options)
+{
+    ForEachItemIn(i, readers)
+    {
+        IDiskRowReader & cur = readers.item(i);
+        if (cur.matches(format, streamRemote, expectedCrc, expected, projectedCrc, projected, actualCrc, actual, options))
+            return &cur;
+    }
+    IDiskRowReader * reader = createDiskReader(format, streamRemote, expectedCrc, expected, projectedCrc, projected, actualCrc, actual, options);
+    readers.append(*reader);
+    return reader;
+}
+
+bool CHThorNewDiskReadBaseActivity::openFilePart(const char * filename)
+{
+    const char * format = "thor";   // more - should extract from the current file (could even mix flat and csv...)
```

Comment:
more?

Reply:
At the moment this activity is only used for binary reads - in the future the same activity will be used for binary/csv/xml reads